drivers/fpga/fpga_altera.c

```diff
@@ -0,0 +1,429 @@
+		return -EINVAL;
+	}
+
+	/* Verify the FPGA config status response */
+	ret = fpga_reconfig_status_validate(
+		(struct fpga_config_status *)priv_data.mbox_response_data);
+
+	k_free(priv_data.mbox_response_data);
+
+	return ret;
+}
+
+static int32_t socfpga_bridges_reset(uint32_t enable)
+{
+	uint32_t smc_cmd[2] = {0};
+	int ret = 0;
+	struct private_data_t priv_data;
+
+	/* Initialize the semaphore */
+	k_sem_init(&(priv_data.smc_sem), 0, 1);
+
+	smc_cmd[SMC_REQUEST_A2_INDEX] = FIELD_GET(BIT(0), enable);
+		
+	smc_cmd[SMC_REQUEST_A2_INDEX] |= BIT(1);
+	smc_cmd[SMC_REQUEST_A3_INDEX] = BRIDGE_MASK;
+	
+	ret = smc_send(SIP_SVC_PROTO_CMD_SYNC, SMC_FUNC_ID_SET_HPS_BRIDGES, smc_cmd, &priv_data);
+	if (ret) {
+		LOG_ERR("Failed to send the smc Command !!");
+		return ret;
+	}
+
+	/* Wait for the SiP SVC callback */
+	k_sem_take(&(priv_data.smc_sem), K_FOREVER);
+
+	/* Check error code */
+	if (priv_data.response.a0) {
+		ret = -ENOMSG;
+	}
+
+	return ret;
+}
+
+static int altera_fpga_on(const struct device *dev)
+{
+    int32_t ret = 0;
+
+	/* Opening SIP SVC session */
+	ret = svc_client_open();
+	if (ret) {
+		LOG_ERR("Client open Failed!");
+		return ret;
+	}
+
+	/* Check FPGA status before bridge enable/disable */
+	ret = is_fpga_config_not_ready();
+	if (ret) {
+		LOG_ERR("FPGA not ready. Bridge reset aborted!");
+		svc_client_close();
+		return -EIO;
+	}
+
+	/* Bridge reset */
+	ret = socfpga_bridges_reset(0x01);
+	if (ret) {
+		LOG_ERR("Bridge reset failed");
+	}
+
+	/* Ignoring the return value to return bridge reset status */
+	if (svc_client_close()) {
+		LOG_ERR("Unregistering & Closing failed");
+	}
+
+	return ret;
+}
+
+static int altera_fpga_off(const struct device *dev)
+{
+	int32_t ret = 0;
+
+	/* Opening SIP SVC session */
+	ret = svc_client_open();
+	if (ret) {
+		LOG_ERR("Client open Failed!");
+		return ret;
+	}
+
+	/* Check FPGA status before bridge enable/disable */
+	ret = is_fpga_config_not_ready();
+	if (ret) {
+		LOG_ERR("FPGA not ready. Bridge reset aborted!");
+		svc_client_close();
+		return -EIO;
+	}
+
+	/* Bridge reset */
+	ret = socfpga_bridges_reset(0x00);
+	if (ret) {
+		LOG_ERR("Bridge reset failed");
+	}
+
+	/* Ignoring the return value to return bridge reset status */
+	if (svc_client_close()) {
+		LOG_ERR("Unregistering & Closing failed");
+	}
+
+	return ret;
+}
+
+static int altera_fpga_init(const struct device *dev)
+{
+    if (!dev) {
+        LOG_ERR("No such device found");
+        return -ENODEV;
+	}
+
+    mailbox_smc_dev = sip_svc_get_controller("smc");
+	if (!mailbox_smc_dev) {
+		LOG_ERR("Arm SiP service not found");
+		return -ENODEV;
+	}
+
+	mailbox_client_token = sip_svc_register(mailbox_smc_dev, NULL);
+	if (mailbox_client_token == SIP_SVC_ID_INVALID) {
+		mailbox_smc_dev = NULL;
+		LOG_ERR("Mailbox client register fail");
+		return -EINVAL;
+	}
+	
+    return 0;
+}
+
+static const struct fpga_driver_api altera_fpga_api = {
+	.on = altera_fpga_on,
+	.off = altera_fpga_off
+};
```

Comment:
Unfortunately the subsystem currently requires the remaining callbacks to be defined, so that cannot go in like this.
At the very least we'd need to modify the common layer to allow NULLs for the rest and return -ENOTSUP. Otherwise you might see random issues when sb tries to use your driver

Reply:
done